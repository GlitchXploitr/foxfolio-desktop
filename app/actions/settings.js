// @flow
import type { SettingsType } from '../reducers/settings';
import type { Action } from './action.d';

export function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}
