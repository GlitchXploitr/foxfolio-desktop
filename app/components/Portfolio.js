// @flow
import React, { Component } from 'react';
import { Grid, Paper } from 'material-ui';
import type { Transaction } from '../reducers/transactions';

type Props = {
  transactions: Transaction[],
  ticker: Object
};

export default class Portfolio extends Component<Props> {

  render() {
    const portfolio = calculatePortfolio(this.props.transactions);
    const ticker = this.props.ticker;

    const keys = Object.keys(portfolio);
    return (
      <Paper style={{ padding: 10 }}>
        <h1>Portfolio</h1>
        <Grid container spacing={24} style={{ padding: 10 }}>
          {keys
            .filter(asset => portfolio[asset] !== 0)
            .map(asset =>
              [
                <Grid item xs={6} md={1}><h2>{asset}</h2></Grid>,
                <Grid item xs={6} md={1}><h3>{portfolio[asset].toFixed(6)}</h3>
                </Grid>,
                <Grid item xs={6} md={2}><h3>= {ticker[asset]
                  ? (parseFloat(ticker[asset].price_eur) * portfolio[asset]).toFixed(2)
                  : 0} â‚¬</h3>
                </Grid>,
              ],
            )}
        </Grid>
      </Paper>
    );
  }
}

function calculatePortfolio(transactions: Transaction[]) {
  return transactions.reduce((acc, transaction) => {
    switch (transaction.type) {
      case 'DEPOSIT':
      case 'WITHDRAW':
        acc[transaction.currency] = acc[transaction.currency] || 0;
        acc[transaction.currency] += transaction.amount;
        break;
      case 'BUY':
        acc[transaction.market.major] = acc[transaction.market.major] || 0;
        acc[transaction.market.minor] = acc[transaction.market.minor] || 0;

        acc[transaction.market.major] -= (transaction.rate * transaction.amount) + transaction.commission;
        acc[transaction.market.minor] += transaction.amount;
        break;
      case 'SELL':
        acc[transaction.market.major] = acc[transaction.market.major] || 0;
        acc[transaction.market.minor] = acc[transaction.market.minor] || 0;

        acc[transaction.market.major] += (transaction.rate * transaction.amount) - transaction.commission;
        acc[transaction.market.minor] -= transaction.amount;
        break;
      default:
        break;
    }
    return acc;
  }, {});
}
