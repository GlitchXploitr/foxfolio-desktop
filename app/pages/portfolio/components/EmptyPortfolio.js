// @flow
import React from 'react';
import { Button, Typography } from 'material-ui';
import { Link } from 'react-router-dom';

export default function EmptyPortfolio() {
  return (
    <div>
      <Typography type="title">No data yet</Typography>
      <Typography type="subheading">
        Try to add an<Button dense color="primary" component={Link} to="/sources">exchange</Button>
        or a<Button dense color="primary" component={Link} to="/wallets">wallet</Button>
      </Typography>
    </div>
  );
}
