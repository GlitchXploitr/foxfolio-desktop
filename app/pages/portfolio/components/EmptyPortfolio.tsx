import { Button, Typography } from 'material-ui';
import React from 'react';
import { Link } from 'react-router-dom';

export const EmptyPortfolio = () => {
  return (
    <div>
      <Typography variant="title">No data yet</Typography>
      <Typography variant="subheading">
        Try to add an<Button
          size="small"
          color="primary"
          component={props => <Link {...props} to="/sources" />}
        >
          exchange
        </Button>
        or a<Button
          size="small"
          color="primary"
          component={props => <Link {...props} to="/wallets" />}
        >
          wallet
        </Button>
      </Typography>
    </div>
  );
};
