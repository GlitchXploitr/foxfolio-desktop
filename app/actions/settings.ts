import { SettingsType } from '../reducers/settings';
import { Action, Dispatch, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}

export function saveSettingsAndUpdateTicker(settings: SettingsType): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(saveSettings(settings));
    dispatch(requestTickerUpdate([], settings.fiatCurrency, settings.cryptoCurrency));
  };
}
