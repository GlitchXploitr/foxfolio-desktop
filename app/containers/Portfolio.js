// @flow
import React from 'react';
import R from 'ramda';
import { Paper } from 'material-ui';
import type { Transaction } from '../actions/transaction.d';
import PortfolioChart from '../components/Portfolio/PortfolioChart';
import type { SettingsType } from '../reducers/settings';
import PortfolioPositions from '../components/Portfolio/PortfolioPositions';
import PortfolioHeader from '../components/Portfolio/PortfolioHeader';
import EmptyPortfolio from '../components/Portfolio/EmptyPortfolio';
import type { Wallet } from '../actions/wallet.d';
import type { Balances, Portfolio } from '../types/portfolio.d.ts';

type Props = {
  balances: { [string]: Balances },
  ticker: Object,
  coinlist: Object,
  wallets: Wallet[],
  settings: SettingsType
};

export default function PortfolioContainer(
  { balances, ticker, coinlist, wallets, settings }: Props) {
  const portfolio = calculatePortfolio(wallets, balances);
  const sum = {
    btc: calculateSum(ticker, portfolio.total, 'BTC'),
    fiat: calculateSum(ticker, portfolio.total, settings.fiatCurrency),
  };

  if (ticker.BTC || ticker.ETH) { // TODO What if the user has no BTC?
    return (
      <div>
        <Paper style={{ marginTop: 0, paddingBottom: 25, paddingTop: 25, textAlign: 'center' }}>
          <PortfolioHeader fiatCurrency={settings.fiatCurrency} sum={sum} ticker={ticker}/>
        </Paper>
        <Paper style={{ marginTop: 30, paddingBottom: 20, paddingTop: 10 }}>
          <PortfolioChart ticker={ticker} portfolio={portfolio.total} sum={sum.btc}/>
        </Paper>
        <Paper style={{ marginTop: 30 }}>
          <PortfolioPositions
            portfolio={portfolio}
            ticker={ticker}
            coinlist={coinlist}
            fiatCurrency={settings.fiatCurrency}
            sumBTC={sum.btc}
          />
        </Paper>
      </div>
    );
  }
  return (
    <Paper style={{ marginTop: 0, paddingBottom: 25, paddingTop: 25, textAlign: 'center' }}>
      <EmptyPortfolio/>
    </Paper>
  );
}

function calculatePortfolio(wallets: Wallet[], balances: { [string]: Balances }): Portfolio {
  const walletBalances = wallets.reduce((acc, wallet) => ({
    ...acc,
    [wallet.currency]: (acc[wallet.currency] || 0) + wallet.quantity,
  }), {});
  let exchangeBalances = {};
  Object.keys(balances).forEach(exchange => {
    exchangeBalances = R.mergeWith((a, b) => a + b, exchangeBalances, balances[exchange]);
  });

  return {
    total: R.mergeWith((a, b) => a + b, exchangeBalances, walletBalances),
    exchanges: balances,
    wallets: walletBalances,
  };
}

function calculateSum(ticker: Object, portfolio: Object, currency: string) {
  return Object.keys(portfolio)
    .filter(asset => ticker[asset])
    .reduce((acc, asset) => acc + (ticker[asset][currency.toUpperCase()].PRICE * portfolio[asset]), 0);
}
