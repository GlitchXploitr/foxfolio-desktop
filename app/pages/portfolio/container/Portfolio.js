// @flow
import React from 'react';
import type { Node } from 'react';
import R from 'ramda';
import { Paper } from 'material-ui';

import PortfolioChart from '../components/PortfolioChart';
import { getTickerPrice } from '../../../helpers/transactions';
import type { Coinlist } from '../../../reducers/coinlist/types.d';
import type { SettingsType } from '../../../reducers/settings';
import PortfolioPositions from '../components/PortfolioPositions';
import PortfolioHeader from '../components/PortfolioHeader';
import EmptyPortfolio from '../components/EmptyPortfolio';
import type { Wallet } from '../../../reducers/wallets/types.d';
import type { Ticker } from '../../../reducers/ticker/types.d';
import type { Balances, Portfolio } from '../types/portfolio.d';

type Props = {
  balances: { [string]: Balances },
  ticker: Ticker,
  coinlist: Coinlist,
  wallets: Wallet[],
  settings: SettingsType
};

export default function PortfolioContainer(
  { balances, ticker, coinlist, wallets, settings }: Props): Node {
  const portfolio = calculatePortfolio(wallets, balances, settings);
  const sum = {
    btc: calculateSum(ticker, portfolio.total, 'BTC'),
    fiat: calculateSum(ticker, portfolio.total, settings.fiatCurrency),
  };
  const change = {
    btc: calculateChange(ticker, portfolio.total, sum.btc, 'BTC'),
    fiat: calculateChange(ticker, portfolio.total, sum.fiat, settings.fiatCurrency),
  };

  if (ticker.BTC || ticker.ETH) { // TODO What if the user has no BTC?
    return (
      <div>
        <Paper style={{ marginTop: 0, paddingBottom: 25, paddingTop: 25, textAlign: 'center' }}>
          <PortfolioHeader change={change} fiatCurrency={settings.fiatCurrency} sum={sum} ticker={ticker}/>
        </Paper>
        <Paper style={{ marginTop: 30, paddingBottom: 20, paddingTop: 10 }}>
          <PortfolioChart ticker={ticker} portfolio={portfolio.total} sum={sum.btc}/>
        </Paper>
        <Paper style={{ marginTop: 30 }}>
          <PortfolioPositions
            portfolio={portfolio}
            ticker={ticker}
            coinlist={coinlist}
            settings={settings}
            sumBTC={sum.btc}
          />
        </Paper>
      </div>
    );
  }
  return (
    <Paper style={{ marginTop: 0, paddingBottom: 25, paddingTop: 25, textAlign: 'center' }}>
      <EmptyPortfolio/>
    </Paper>
  );
}

function calculatePortfolio(wallets: Wallet[], balances: { [string]: Balances }, settings: SettingsType): Portfolio {
  const walletBalances = wallets
    .filter(wallet => !(wallet.currency === settings.fiatCurrency && settings.includeFiat))
    .reduce((acc, wallet) => ({
      ...acc,
      [wallet.currency]: (acc[wallet.currency] || 0) + wallet.quantity,
    }), {});

  const filteredBalances = R.map(settings.includeFiat ? R.identity : R.omit([settings.fiatCurrency]))(balances);
  let exchangeBalances = {};
  Object.keys(filteredBalances).forEach(exchange => {
    exchangeBalances = R.mergeWith((a, b) => a + b, exchangeBalances, filteredBalances[exchange]);
  });

  return {
    total: R.mergeWith((a, b) => a + b, exchangeBalances, walletBalances),
    exchanges: filteredBalances,
    wallets: walletBalances,
  };
}

function calculateSum(ticker: Ticker, portfolio: Object, currency: string) {
  return Object.keys(portfolio)
    .filter(asset => ticker[asset])
    .reduce((acc, asset) => acc + (getTickerPrice(ticker, asset, currency) * portfolio[asset]), 0);
}

function calculateChange(ticker: Ticker, portfolio: Object, sum: number, currency: string) {
  return Object.keys(portfolio)
    .filter(asset => ticker[asset])
    .filter(asset => asset !== currency)
    .reduce(
      (acc, asset) =>
        acc + (
          ticker[asset][currency].CHANGEPCT24HOUR
          * ((getTickerPrice(ticker, asset, currency) * portfolio[asset]) / sum)
        ),
      0);
}
