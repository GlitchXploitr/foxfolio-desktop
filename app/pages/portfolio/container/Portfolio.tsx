import * as _ from 'lodash';
import { Button, Paper } from 'material-ui';
import React from 'react';
import { Coinlist } from '../../../reducers/coinlist';
import { SettingsType } from '../../../reducers/settings';
import { Ticker } from '../../../reducers/ticker';
import { EmptyPortfolio } from '../components/EmptyPortfolio';
import { PortfolioChart } from '../components/PortfolioChart';
import { PortfolioFocusButtons } from '../components/PortfolioFocusButtons';
import { PortfolioHeader } from '../components/PortfolioHeader';
import { PortfolioPositions } from '../components/PortfolioPositions';
import { Portfolio, PortfolioChange, PortfolioSum } from '../types/portfolio.types';

export interface PortfolioProps {
  coinlist: Coinlist;
  ticker: Ticker;
  portfolio: Portfolio;
  sum: PortfolioSum;
  change: PortfolioChange;
  settings: SettingsType;
}

export const PortfolioContainer = ({
  portfolio,
  ticker,
  coinlist,
  sum,
  change,
  settings,
}: PortfolioProps) => {
  if (ticker[settings.cryptoCurrency] && !_.isEmpty(portfolio.total)) {
    return (
      <div>
        <Paper style={{ marginTop: 0, paddingBottom: 25, paddingTop: 25, textAlign: 'center' }}>
          <PortfolioHeader change={change} settings={settings} sum={sum} />
        </Paper>
        <Paper style={{ marginTop: 30, paddingBottom: 20, paddingTop: 10 }}>
          <PortfolioChart
            ticker={ticker}
            portfolio={portfolio.total}
            sum={sum.crypto}
            settings={settings}
          />
        </Paper>
        <div style={{ marginTop: 30 }}>
          <Paper square style={{ width: 192 }}>
            <PortfolioFocusButtons />
          </Paper>
        </div>
        <div>
          <PortfolioPositions
            portfolio={portfolio}
            ticker={ticker}
            coinlist={coinlist}
            settings={settings}
          />
        </div>
      </div>
    );
  }
  return (
    <Paper style={{ marginTop: 0, paddingBottom: 25, paddingTop: 25, textAlign: 'center' }}>
      <EmptyPortfolio />
    </Paper>
  );
};
