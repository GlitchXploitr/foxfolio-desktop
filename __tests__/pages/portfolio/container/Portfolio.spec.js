// @flow
import * as React from 'react';
import { createShallow } from 'material-ui/test-utils';
import EmptyPortfolio from '../../../../app/pages/portfolio/components/EmptyPortfolio';
import Portfolio from '../../../../app/pages/portfolio/container/Portfolio';
import type { Balances } from '../../../../app/reducers/exchanges/types.d';
import { initialSettings } from '../../../../app/reducers/settings';
import type { Wallet } from '../../../../app/reducers/wallets/types.d';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

let balances: { [string]: Balances };
let coinlist;
let ticker;
let wallets: Wallet[];

beforeEach(() => {
  balances = {};
  coinlist = {};
  ticker = {};
  wallets = [];
});


test('Render portfolio page', () => {
  balances = { exchange: { BTC: 5 } };
  ticker = { BTC: { USD: { PRICE: 1000, CHANGEPCT24HOUR: 2 } } };
  wallets = [{ address: '', currency: 'BTC', quantity: 1 }];

  const wrapper = shallow(
    <Portfolio
      balances={balances}
      coinlist={coinlist}
      settings={initialSettings}
      ticker={ticker}
      wallets={wallets}
    />);
  expect(wrapper).toMatchSnapshot();
});

test('Render empty portfolio when ticker is empty', () => {
  balances = { exchange: { BTC: 1 } };

  const wrapper = shallow(
    <Portfolio
      balances={balances}
      coinlist={coinlist}
      settings={initialSettings}
      ticker={ticker}
      wallets={wallets}
    />);
  expect(wrapper.contains(<EmptyPortfolio/>)).toBeTruthy();
});

test('Render empty portfolio when balances and wallets are empty', () => {
  ticker = { BTC: { USD: { PRICE: 1000, CHANGEPCT24HOUR: 2 } } };

  const wrapper = shallow(
    <Portfolio
      balances={balances}
      coinlist={coinlist}
      settings={initialSettings}
      ticker={ticker}
      wallets={wallets}
    />);
  expect(wrapper.contains(<EmptyPortfolio/>)).toBeTruthy();
});

