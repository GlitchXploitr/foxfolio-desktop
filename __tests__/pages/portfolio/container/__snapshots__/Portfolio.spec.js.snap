// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render portfolio page 1`] = `
<div>
  <WithStyles(Paper)
    style={
      Object {
        "marginTop": 0,
        "paddingBottom": 25,
        "paddingTop": 25,
        "textAlign": "center",
      }
    }
  >
    <PortfolioHeader
      change={
        Object {
          "crypto": 0,
          "fiat": 2,
        }
      }
      settings={
        Object {
          "cryptoCurrency": "BTC",
          "currencyFocus": "fiat",
          "fiatCurrency": "USD",
          "includeFiat": true,
          "theme": "light",
        }
      }
      sum={
        Object {
          "crypto": 6,
          "fiat": 6000,
        }
      }
    />
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "marginTop": 30,
        "paddingBottom": 20,
        "paddingTop": 10,
      }
    }
  >
    <WithTheme(PortfolioChart)
      portfolio={
        Object {
          "BTC": 6,
        }
      }
      settings={
        Object {
          "cryptoCurrency": "BTC",
          "currencyFocus": "fiat",
          "fiatCurrency": "USD",
          "includeFiat": true,
          "theme": "light",
        }
      }
      sum={6}
      ticker={
        Object {
          "BTC": Object {
            "USD": Object {
              "CHANGEPCT24HOUR": 2,
              "PRICE": 1000,
            },
          },
        }
      }
    />
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    <PortfolioPositions
      coinlist={Object {}}
      portfolio={
        Object {
          "exchanges": Object {
            "exchange": Object {
              "BTC": 5,
            },
          },
          "total": Object {
            "BTC": 6,
          },
          "wallets": Object {
            "BTC": 1,
          },
        }
      }
      settings={
        Object {
          "cryptoCurrency": "BTC",
          "currencyFocus": "fiat",
          "fiatCurrency": "USD",
          "includeFiat": true,
          "theme": "light",
        }
      }
      ticker={
        Object {
          "BTC": Object {
            "USD": Object {
              "CHANGEPCT24HOUR": 2,
              "PRICE": 1000,
            },
          },
        }
      }
    />
  </WithStyles(Paper)>
</div>
`;
